// playwright-pdf-reporter/dist/options.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.resolveOptions = void 0;
const path_1 = __importDefault(require("path"));
const defaultOptions = {
    outputDir: 'playwright-report/pdf',
    fileName: 'test-report.pdf',
    theme: 'light',
    includeScreenshots: true,
    includeHtml: true,
    reportType: 'full',
    historicalDataPath: 'playwright-report/pdf/history.json',
    trendLabel: 'Overall',
    metadata: {},
    scope: {},
    customSections: {},
    bugTrackerBaseUrl: undefined,
    fastMode: false
};
const resolveOptions = (options) => {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const merged = {
        ...defaultOptions,
        ...options,
        metadata: {
            ...defaultOptions.metadata,
            ...options === null || options === void 0 ? void 0 : options.metadata
        },
        scope: {
            ...defaultOptions.scope,
            ...options === null || options === void 0 ? void 0 : options.scope
        },
        customSections: {
            ...defaultOptions.customSections,
            ...options === null || options === void 0 ? void 0 : options.customSections
        }
    };
    merged.outputDir = path_1.default.normalize(merged.outputDir || defaultOptions.outputDir);
    const historyOptionProvided = options && Object.prototype.hasOwnProperty.call(options, 'historicalDataPath');
    if (historyOptionProvided) {
        // Explicitly provided by the user:
        // - non-empty string → normalize and use
        // - empty string / undefined → disable history
        merged.historicalDataPath = merged.historicalDataPath
            ? path_1.default.normalize(merged.historicalDataPath)
            : undefined;
    }
    else if (defaultOptions.historicalDataPath) {
        // No value provided → fall back to default
        merged.historicalDataPath = path_1.default.normalize(defaultOptions.historicalDataPath);
    }
    if (!((_a = merged.fileName) === null || _a === void 0 ? void 0 : _a.toLowerCase().endsWith('.pdf'))) {
        merged.fileName = `${((_b = merged.fileName) !== null && _b !== void 0 ? _b : defaultOptions.fileName).replace(/\.pdf$/i, '')}.pdf`;
    }
    merged.reportType = (_c = merged.reportType) !== null && _c !== void 0 ? _c : defaultOptions.reportType;
    merged.theme = (_d = merged.theme) !== null && _d !== void 0 ? _d : defaultOptions.theme;
    merged.includeScreenshots = (_e = merged.includeScreenshots) !== null && _e !== void 0 ? _e : defaultOptions.includeScreenshots;
    merged.includeHtml = (_f = merged.includeHtml) !== null && _f !== void 0 ? _f : defaultOptions.includeHtml;
    merged.trendLabel = (_g = merged.trendLabel) !== null && _g !== void 0 ? _g : defaultOptions.trendLabel;
    merged.fastMode = (_h = merged.fastMode) !== null && _h !== void 0 ? _h : defaultOptions.fastMode;
    if (merged.fastMode) {
        // Fast mode trades some richness for speed.
        merged.includeScreenshots = false;
        merged.includeHtml = false;
        merged.historicalDataPath = undefined;
    }
    return merged;
};
exports.resolveOptions = resolveOptions;
