// playwright-pdf-reporter/dist/pdfGenerator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.persistHtmlArtifact = exports.htmlToPdf = void 0;
const path_1 = __importDefault(require("path"));
const playwright_core_1 = require("playwright-core");
const fs_1 = require("./utils/fs");
const htmlToPdf = async (html, pdfPath, metadata) => {
    await (0, fs_1.ensureDir)(path_1.default.dirname(pdfPath));
    const browser = await playwright_core_1.chromium.launch({
        headless: true
    });
    try {
        const page = await browser.newPage();
        await page.setContent(html, { waitUntil: 'networkidle' });
        await page.addStyleTag({
            content: `
        @page {
          size: A4;
          margin: 24px;
        }
      `
        });
        await page.evaluate(({ title, author }) => {
            document.title = title;
            document.documentElement.setAttribute('data-author', author);
        }, { title: metadata.title, author: metadata.author });
        // Give charts/layout a brief moment to settle without adding a large delay.
        await page.waitForTimeout(100);
        await page.pdf({
            path: pdfPath,
            format: 'A4',
            printBackground: true,
            displayHeaderFooter: false
        });
    }
    finally {
        await browser.close();
    }
};
exports.htmlToPdf = htmlToPdf;
const persistHtmlArtifact = async (html, filePath) => {
    await (0, fs_1.writeFileSafe)(filePath, html, 'utf-8');
};
exports.persistHtmlArtifact = persistHtmlArtifact;
