// playwright-pdf-reporter/dist/reportBuilder.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildReportData = void 0;
const format_1 = require("./utils/format");
const package_json_1 = __importDefault(require("../package.json"));
const buildReportData = (cases, config, options, warnings, historyEntries) => {
    const summary = computeSummary(cases);
    const metadata = resolveMetadata(options);
    const scope = resolveScope(options);
    const sections = resolveSections(options);
    const environment = buildEnvironment(config);
    const failures = cases.filter((item) => item.status !== 'passed');
    const metrics = computeMetrics(summary);
    const history = historyEntries.slice(-20);
    return {
        options,
        summary,
        scope,
        metadata,
        environment,
        cases,
        failures,
        metrics,
        history,
        sections,
        warnings,
        reporterVersion: package_json_1.default.version
    };
};
exports.buildReportData = buildReportData;
const computeSummary = (cases) => {
    const total = cases.length;
    const isFlaky = (testCase) => {
        if (testCase.annotations['flaky']) {
            return true;
        }
        if (!testCase.attempts || testCase.attempts.length < 2) {
            return false;
        }
        const latest = testCase.attempts[testCase.attempts.length - 1];
        const hadFailure = testCase.attempts.some((attempt) => attempt.status === 'failed');
        return (latest === null || latest === void 0 ? void 0 : latest.status) === 'passed' && hadFailure;
    };
    const flakyCases = cases.filter(isFlaky);
    const flakyIds = new Set(flakyCases.map((c) => c.id));
    const passed = cases.filter((c) => c.status === 'passed').length;
    const failed = cases.filter((c) => !flakyIds.has(c.id) && (c.status === 'failed' || c.status === 'timedOut')).length;
    const skipped = cases.filter((c) => c.status === 'skipped').length;
    const flaky = flakyCases.length;
    const durationMs = cases.reduce((acc, c) => acc + (c.duration || 0), 0);
    const startTime = Math.min(...cases.map((c) => { var _a; return (_a = c.startedAt) !== null && _a !== void 0 ? _a : Date.now(); }));
    const endTime = Math.max(...cases.map((c) => { var _a; return (_a = c.completedAt) !== null && _a !== void 0 ? _a : Date.now(); }));
    return {
        total,
        passed,
        failed,
        skipped,
        flaky,
        durationMs,
        startTime: Number.isFinite(startTime) ? startTime : Date.now(),
        endTime: Number.isFinite(endTime) ? endTime : Date.now()
    };
};
const computeMetrics = (summary) => {
    const executed = summary.total - summary.skipped;
    const coveragePercent = (0, format_1.clamp)((0, format_1.percent)(executed, summary.total || 1));
    const reliabilityScore = (0, format_1.clamp)((0, format_1.percent)(summary.passed, executed || 1));
    const maintainabilityIndex = (0, format_1.clamp)(100 - summary.failed * 5, 40, 100);
    const reusabilityScore = (0, format_1.clamp)(70 + summary.passed * 2 - summary.failed * 3, 40, 100);
    return {
        coveragePercent,
        reliabilityScore,
        maintainabilityIndex,
        reusabilityScore
    };
};
const resolveMetadata = (options) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k, _l, _m, _o, _p, _q, _r, _s, _t, _u, _v, _w;
    return ({
        title: (_b = (_a = options.metadata) === null || _a === void 0 ? void 0 : _a.title) !== null && _b !== void 0 ? _b : 'Playwright Automation Report',
        author: (_e = (_d = (_c = options.metadata) === null || _c === void 0 ? void 0 : _c.author) !== null && _d !== void 0 ? _d : process.env.GIT_AUTHOR_NAME) !== null && _e !== void 0 ? _e : 'Automation Bot',
        build: (_h = (_g = (_f = options.metadata) === null || _f === void 0 ? void 0 : _f.build) !== null && _g !== void 0 ? _g : process.env.BUILD_ID) !== null && _h !== void 0 ? _h : 'local',
        environment: (_l = (_k = (_j = options.metadata) === null || _j === void 0 ? void 0 : _j.environment) !== null && _k !== void 0 ? _k : process.env.NODE_ENV) !== null && _l !== void 0 ? _l : 'local',
        project: (_o = (_m = options.metadata) === null || _m === void 0 ? void 0 : _m.project) !== null && _o !== void 0 ? _o : 'default',
        release: (_r = (_q = (_p = options.metadata) === null || _p === void 0 ? void 0 : _p.release) !== null && _q !== void 0 ? _q : process.env.RELEASE_NAME) !== null && _r !== void 0 ? _r : 'rolling',
        ciLink: (_u = (_t = (_s = options.metadata) === null || _s === void 0 ? void 0 : _s.ciLink) !== null && _t !== void 0 ? _t : process.env.CI_JOB_URL) !== null && _u !== void 0 ? _u : '',
        tags: (_w = (_v = options.metadata) === null || _v === void 0 ? void 0 : _v.tags) !== null && _w !== void 0 ? _w : []
    });
};
const resolveScope = (options) => {
    var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
    return ({
        objectives: (_b = (_a = options.scope) === null || _a === void 0 ? void 0 : _a.objectives) !== null && _b !== void 0 ? _b : [
            'Validate end-to-end user journeys',
            'Ensure regression stability for critical flows'
        ],
        dataSets: (_d = (_c = options.scope) === null || _c === void 0 ? void 0 : _c.dataSets) !== null && _d !== void 0 ? _d : ['Synthetic test data', 'Seeded accounts'],
        passCriteria: (_f = (_e = options.scope) === null || _e === void 0 ? void 0 : _e.passCriteria) !== null && _f !== void 0 ? _f : [
            'All P0/P1 cases pass',
            'No critical regressions filed'
        ],
        risks: (_h = (_g = options.scope) === null || _g === void 0 ? void 0 : _g.risks) !== null && _h !== void 0 ? _h : ['Flaky network environments', '3rd party dependencies'],
        alignment: (_k = (_j = options.scope) === null || _j === void 0 ? void 0 : _j.alignment) !== null && _k !== void 0 ? _k : 'Supports release readiness and CI quality gates'
    });
};
const resolveSections = (options) => {
    var _a, _b, _c, _d, _e, _f;
    return ({
        challenges: (_b = (_a = options.customSections) === null || _a === void 0 ? void 0 : _a.challenges) !== null && _b !== void 0 ? _b : [
            'Intermittent latency from downstream APIs',
            'Maintenance of shared fixtures across suites'
        ],
        lessonsLearned: (_d = (_c = options.customSections) === null || _c === void 0 ? void 0 : _c.lessonsLearned) !== null && _d !== void 0 ? _d : [
            'Parallelizing specs reduced suite time by 30%',
            'Centralized test data catalog improved reusability'
        ],
        recommendations: (_f = (_e = options.customSections) === null || _e === void 0 ? void 0 : _e.recommendations) !== null && _f !== void 0 ? _f : [
            'Stabilize flaky selectors using resilient locating strategies',
            'Automate build health checks with PDF summaries in CI'
        ]
    });
};
const buildEnvironment = (config) => {
    var _a, _b, _c, _d;
    if (!config) {
        return {
            browsers: ['chromium'],
            devices: [],
            operatingSystems: [process.platform],
            configMatrix: []
        };
    }
    const browsers = new Set();
    const devices = new Set();
    const osVersions = new Set();
    const configMatrix = [];
    for (const project of (_a = config.projects) !== null && _a !== void 0 ? _a : []) {
        const use = project.use;
        if (typeof use.browserName === 'string') {
            browsers.add(use.browserName);
        }
        if (typeof use.channel === 'string') {
            browsers.add(`${(_b = use.browserName) !== null && _b !== void 0 ? _b : 'chromium'} (${use.channel})`);
        }
        if (typeof use.device === 'string') {
            devices.add(use.device);
        }
        if (typeof use.viewport === 'object' && use.viewport) {
            configMatrix.push({
                project: project.name,
                viewport: `${use.viewport.width}x${use.viewport.height}`,
                headless: `${(_c = use.headless) !== null && _c !== void 0 ? _c : true}`
            });
        }
        else {
            configMatrix.push({
                project: project.name,
                headless: `${(_d = use.headless) !== null && _d !== void 0 ? _d : true}`
            });
        }
        if (typeof use.platform === 'string') {
            osVersions.add(use.platform);
        }
    }
    if (!osVersions.size) {
        osVersions.add(process.platform);
    }
    return {
        browsers: Array.from(browsers),
        devices: Array.from(devices),
        operatingSystems: Array.from(osVersions),
        configMatrix
    };
};
