// playwright-pdf-reporter/dist/utils/format.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.slugify = exports.formatNumber = exports.toDateTimeString = exports.clamp = exports.percent = exports.formatDuration = void 0;
const formatDuration = (ms) => {
    if (!ms || Number.isNaN(ms)) {
        return '0s';
    }
    const seconds = Math.floor(ms / 1000);
    const mins = Math.floor(seconds / 60);
    const hrs = Math.floor(mins / 60);
    const remainingSeconds = seconds % 60;
    const remainingMinutes = mins % 60;
    const parts = [];
    if (hrs)
        parts.push(`${hrs}h`);
    if (remainingMinutes)
        parts.push(`${remainingMinutes}m`);
    if (remainingSeconds || parts.length === 0)
        parts.push(`${remainingSeconds}s`);
    return parts.join(' ');
};
exports.formatDuration = formatDuration;
const percent = (value, total) => {
    if (!total)
        return 0;
    return Math.round((value / total) * 10000) / 100;
};
exports.percent = percent;
const clamp = (value, min = 0, max = 100) => Math.min(Math.max(value, min), max);
exports.clamp = clamp;
const toDateTimeString = (timestamp) => {
    return new Date(timestamp).toLocaleString();
};
exports.toDateTimeString = toDateTimeString;
const formatNumber = (value) => Intl.NumberFormat().format(value);
exports.formatNumber = formatNumber;
const slugify = (value) => value
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/(^-|-$)/g, '');
exports.slugify = slugify;
